ls1intum/Artemis
# [Bugfix] Student can see feedback before assessment due date

### Checklist

### Motivation and Context


Apparently in some cases can a Student see the manual feedback before the assessment due date is over. This might happen if the manual result gets merged with an updated test grading. 
Fixes: #2749 

### Description

The broadcast of results now checkes if the assessment due date is over before we notify the students. The received result is then updated in the current result list instead of just appended.
In addition, before we allow the online editor to be able to show manual feedback for a student, we check if the assessment due date is either null or has already passed. 


### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise and activate the manual assessment 
3.1 set the assessment due date to a date in the near future
3.2 do not set an assessment date but activate it
4. Participate as a student and assess it as a tutor
5. Verifiy that a student can not see the manual feedback before the assessment due date
6. Now change the grading and save it. Then click on re-evaluate and click on trigger all
7. Verify again that a student can not see the manual feedback before the assessment due date
8. Wait after the assess. due date has passed and verify that a student can see the feedback
9. Another test: 
9.1 Open two tabs and be logged in as a student in Artemis and have the build plan open. The student opens the course-exercise view and watches the result history. Now in the build plan manually trigger the build, and wait for it to finish. Before the assessment due date, the result should not be added to the result history. (This is currently the case in develop but fixed with this PR, see screenshot)

### Test Coverage



ResultResource.java: 83%


### Screenshots


#### Visible Feedback after assess. due date (Locked Repo):
![image](https://user-images.githubusercontent.com/33484318/109715828-6720fc00-7ba4-11eb-8eaf-0b6e1ae35aff.png)

#### Not visible feedback before assess. due date (Locked Repo):
![image](https://user-images.githubusercontent.com/33484318/109716057-9e8fa880-7ba4-11eb-89de-9c44482d9a31.png)

#### Wrong received result before the assessment due date:
![image](https://user-images.githubusercontent.com/33484318/110087230-abbeba00-7d93-11eb-9562-c94e50dbbfb3.png)
